src/plugins/intel_cpu/src/nodes/transpose.cpp

```diff
@@ -37,8 +37,56 @@ bool Transpose::isSupportedOperation(const std::shared_ptr<const ov::Node>& op,
     return true;
 }
 
+class TransposeShapeInfer : public ShapeInferEmptyPads {
+public:
+    TransposeShapeInfer(const size_t& out_rank, const std::vector<size_t>& axes_vec)
+    : m_out_rank(out_rank), m_axes_vec(axes_vec), m_outputShape(std::move(VectorDims(out_rank, 1))), m_needReverse(axes_vec.empty()) {}
+
+    std::vector<VectorDims> infer(
+        const std::vector<std::reference_wrapper<const VectorDims>>& input_shapes,
+        const std::unordered_map<size_t, MemoryPtr>& data_dependency) override {
+        const VectorDims& shapeIn = input_shapes[0].get();
+        if (m_needReverse) {
+            for (auto i = 0; i < m_out_rank; ++i) {
+                m_outputShape[i] = shapeIn[m_out_rank - 1 - i];
+            }
+        } else {
+            for (auto i = 0; i < m_out_rank; ++i) {
+                m_outputShape[i] = shapeIn[m_axes_vec[i]];
+            }
+        }
+        return {m_outputShape};
+    }
+
+    port_mask_t get_port_mask() const override {
+        return EMPTY_PORT_MASK;
+    }
+
+private:
+    const size_t m_out_rank;
+    const std::vector<size_t> m_axes_vec;
+    VectorDims m_outputShape;
+    const bool m_needReverse;
+};
+
+class TransposeShapeInferFactory : public ShapeInferFactory {
+public:
+    TransposeShapeInferFactory(const std::shared_ptr<ov::Node>& op) : m_op(op) {}
+    ShapeInferPtr makeShapeInfer() const override {
+        if (const auto order = ov::as_type_ptr<const ov::op::v0::Constant>(m_op->get_input_node_shared_ptr(ov::op::v1::Transpose::ORDER))) {
+            const auto axes_vec = order->cast_vector<size_t>();
+            return std::make_shared<TransposeShapeInfer>(m_op->get_output_partial_shape(0).rank().get_length(), axes_vec);
+        } else {
+            IE_THROW() << "Unexpected operation type in the Transpose shape inference factory";
```

Comment:
I think the error message should be about the fact that the input is not constant.

Reply:
> I think the error message should be about the fact that the input is not constant.

Yes, update the error info.

Reply:
@maxnick @praasz 
Hi all, 
I have something to discuss here.

- The lastest pytorch test hits this bug in the PR https://github.com/openvinotoolkit/openvino/pull/15614. To be honest, former tests don't verify this unusual case because the CI  pass all tests before 15614 merged.
- In the CPU plugin, The second input is always a constant node, which is expected.
- But if the second input is not a constant node, the framework should NOT stop here, right. As we can see https://github.com/openvinotoolkit/openvino/blob/310c4deab9f3428a95347059f144563a1aea2c22/src/plugins/intel_cpu/src/nodes/transpose.cpp#L21, the operation will fall back to the ngraph implementaion to keep the model going.
- As a result, I remove such an **_exception_** here.
looking forward to your comments. thx.

Reply:
If you mean `ngraph` fallback implementation is template implementation for transpose shape inference then:

1. If shape infer on `ov::PartialShape` the result will some dynamic shape.
2. If working with `StaticShapes` then if order is not constant (parameter) and data are passed in tensor map the inference produce output shape, but if 2nd input is Parameter and no data pass it will try create dynamic shape which will trigger to throw exception

So if 2nd input is not constant but there are data which can be pass in tensor map then it should be fine from shape inference perspective, otherwise it will produce dynamic output shape.

Reply:
Thanks for your comments, which helps understand the flow.